### PR TITLE
[#405] trace transformer class

### PR DESCRIPTION
**Pull request by CodiePP**
description
-----------

- [x] `ToLogObject` transforms from a `Tracer m b` to a `Tracer m (LogObject a)` which can be fed into the switchboard. This PR adds a mostly generic transformer `Transformable`. This can also be used to implement specific instances for user-defined types.

checklist
---------

- [x] compiles (`cabal new-clean; cabal new-build`)
- [x] tests run successfully (`cabal new-test`)
- [ ] documentation added and created (`cd docs; nix-shell --run make`)
- [x] link to an issue
- [ ] link to an epic
- [x] add estimate points
- [x] add milestone (the same as the linked issue)

### CODE REVIEW

File: iohk-monitoring/src/Cardano/BM/Data/Tracer.lhs [hunk 1]
```diff
@@ -146,3 +150,80 @@ instance ToJSON a => ToObject (LogObject a)
...
+            LogObject <$> pure ""
+                      <*> (mkLOMeta Debug Public)
+                      <*> pure (LogValue name $ PureD $ realToFrac arg)
+
+instance Transformable a IO Int where
+    trTransformer tr = trFromIntegral tr "int"
+instance Transformable a IO Integer where
+    trTransformer tr = trFromIntegral tr "integer"
+instance Transformable a IO Word64 where
+    trTransformer tr = trFromIntegral tr "integer"
```
Comment (njd42) [inline iohk-monitoring/src/Cardano/BM/Data/Tracer.lhs]: should this not be "word64"

File: iohk-monitoring/src/Cardano/BM/Data/Tracer.lhs [hunk 2]
```diff
@@ -146,3 +150,80 @@ instance ToJSON a => ToObject (LogObject a)
...
+instance Transformable a IO Double where
+    trTransformer tr = trFromReal tr "dbl"
```
Comment (njd42) [inline iohk-monitoring/src/Cardano/BM/Data/Tracer.lhs]: I suggest that "double" and "float" are spelled out in full

File: iohk-monitoring/src/Cardano/BM/Data/Tracer.lhs [hunk 3]
```diff
@@ -146,3 +150,80 @@ instance ToJSON a => ToObject (LogObject a)
...
+instance Transformable a IO Float where
+    trTransformer tr = trFromReal tr "flt"
```
Comment (njd42) [inline iohk-monitoring/src/Cardano/BM/Data/Tracer.lhs]: see above

File: iohk-monitoring/test/Cardano/BM/Test/Structured.lhs [hunk 4]
```diff
@@ -34,15 +38,33 @@ tests = testGroup "Testing Structured Logging" [
...
-    cfg <- defaultConfigTesting
-    baseTrace :: Trace IO Text <- Setup.setupTrace (Right cfg) "logText"
-
-    let logTrace = toLogObject $ baseTrace
+    cfg <- defaultConfigStdout
+    baseTrace :: Tracer IO (LogObject Text) <- Setup.setupTrace (Right cfg) "logText"
 
-    traceWith logTrace "This is a simple message."
-    traceWith logTrace ".. and another!"
+    traceWith (toLogObject baseTrace) ("This is a simple message." :: Text)
```
Comment (njd42) [inline iohk-monitoring/test/Cardano/BM/Test/Structured.lhs]: I"m wondering if there is different way to structure this example that is more in keeping with the expect use, ie describe the example as a function taking a tracer so the lines are
 > traceWith tr ....
and pass in (toLogobject tr 
would this illustrate the abstraction better ? 
People are going to cut and paste the examples - need to make sure that will lead them (generally) into good practices

Comment (CodiePP) [reply iohk-monitoring/test/Cardano/BM/Test/Structured.lhs]: yes, I'll prepare a different one with more transformers and without Text or Pet. Maybe something existing..

**Pull request merged**